rclcpp/src/rclcpp/parameter_events_subscriber.cpp

```diff
@@ -0,0 +1,160 @@
+  qos_(qos)
+{}
+
+void ParameterEventsSubscriber::add_namespace_event_subscriber(const std::string & node_namespace)
+{
+  if (std::find(node_namespaces_.begin(), node_namespaces_.end(),
+    node_namespace) == node_namespaces_.end())
+  {
+    node_namespaces_.push_back(node_namespace);
+    auto topic = join_path(node_namespace, "parameter_events");
+    RCLCPP_DEBUG(node_logging_->get_logger(), "Subscribing to topic: %s", topic.c_str());
+
+    auto event_sub = rclcpp::create_subscription<rcl_interfaces::msg::ParameterEvent>(
+      node_topics_, topic, qos_,
+      std::bind(&ParameterEventsSubscriber::event_callback, this, std::placeholders::_1));
+
+    event_subscriptions_.push_back(event_sub);
+  }
+}
+
+void ParameterEventsSubscriber::set_event_callback(
+  std::function<void(const rcl_interfaces::msg::ParameterEvent::SharedPtr &)> callback,
+  const std::string & node_namespace)
+{
+  std::string full_namespace;
+  if (node_namespace == "") {
+    full_namespace = node_base_->get_namespace();
+  } else {
+    full_namespace = resolve_path(full_namespace);
+  }
+  add_namespace_event_subscriber(full_namespace);
+  user_callback_ = callback;
+}
+
+void ParameterEventsSubscriber::register_parameter_callback(
+  const std::string & parameter_name,
+  std::function<void(const rclcpp::Parameter &)> callback,
+  const std::string & node_name)
+{
+  auto full_node_name = resolve_path(node_name);
+  add_namespace_event_subscriber(split_path(full_node_name).first);
+  parameter_callbacks_[{parameter_name, full_node_name}] = callback;
+}
+
+bool ParameterEventsSubscriber::get_parameter_from_event(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event,
+  rclcpp::Parameter & parameter,
+  const std::string parameter_name,
+  const std::string node_name)
+{
+  if (event->node == resolve_path(node_name)) {
+    rclcpp::ParameterEventsFilter filter(event, {parameter_name},
+      {rclcpp::ParameterEventsFilter::EventType::NEW,
+        rclcpp::ParameterEventsFilter::EventType::CHANGED});
+    if (!filter.get_events().empty()) {
+      const auto & events = filter.get_events();
+      auto param_msg = events[events.size() - 1].second;
+      parameter = rclcpp::Parameter::from_parameter_msg(*param_msg);
+      return true;
+    }
+  }
+  return false;
+}
+
+void ParameterEventsSubscriber::event_callback(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+{
+  std::string node_name = event->node;
+  RCLCPP_DEBUG(node_logging_->get_logger(), "Parameter event received for node: %s",
+    node_name.c_str());
+
+  std::unordered_map<
+    std::pair<std::string, std::string>,
+    std::function<void(const rclcpp::Parameter &)>,
+    stringPairHash>::iterator it;
+
+  for (it = parameter_callbacks_.begin(); it != parameter_callbacks_.end(); ++it) {
+    rclcpp::Parameter p;
+    if (get_parameter_from_event(event, p, it->first.first, it->first.second)) {
+      it->second(p);
+    }
+  }
+
+  if (user_callback_) {
```

Comment:
What if this is set to `nullptr` with `set_event_callback()` after this line, but before the next?